src/connection/server_support.c

```diff
@@ -1289,6 +1289,46 @@ css_start_shutdown_server ()
   css_Server_shutdown_inited = true;
 }
 
+int
+css_register_ha_server (const char *server_name)
+{
+  assert (server_name != NULL);
+  assert (!HA_DISABLED ());
+
+  CSS_CONN_ENTRY *conn;
+  std::string message_to_master = css_pack_message_to_master (server_name);
+
+  // connection is established before server recovery, only when ha_mode is turned on.
+  conn =
+    css_connect_to_master_server (prm_get_integer_value (PRM_ID_TCP_PORT_ID), message_to_master.c_str (),
+				  message_to_master.size ());
+  if (conn != NULL)
+    {
+      int status = hb_register_to_master (conn,
+					  get_server_type () ==
+					  SERVER_TYPE_TRANSACTION ? HB_PTYPE_TRAN_SERVER : HB_PTYPE_PAGE_SERVER);
+
+      if (status != NO_ERROR)
+	{
+	  fprintf (stderr, "failed to hearbeat register to master \n");
+	  css_close_connection_to_master ();
+
+	  return status;
+	}
+      else
+	{
+	  // established connection will be re-used in css_init () after server recovery.
+	  css_Master_conn = conn;
+	  return NO_ERROR;
+	}
+    }
+  else
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ERR_CSS_ERROR_DURING_SERVER_CONNECT, 1, server_name);
```

Comment:
I think the following error message is not enough to describe this un-expected result. It seems like you just follow the error handling in the css_init(). Share your opinion.
> ERR_CSS_ERROR_DURING_SERVER_CONNECT -368 Communication error during connect for %1$s.

Reply:
It is true that it is the same as the error situation in css_init(). 
If an error occurs during the connection to cub_master from a server with a `server_name`, 
the above error message is printed out. 

I agree that the error message itself is not clear in explaining the situation. 

Changing it to 
'Communication error during connect to cub_master from %1$s.' 
would be more appropriate, but it goes beyond the scope of the current issue.

Reply:
The error code ERR_CSS_TCP_CANNOT_CONNECT_TO_MASTER you mentioned in the comment can be set in the callstack you use (css_connect_to_master_server() -> css_common_connect()).

Let's leave it you are written for now.